internal/modules/modules.go

```diff
@@ -10,123 +10,174 @@ package modules
+		if !uriIsLocal(uri) {
+			var err error
+			version, err = getLatestModuleForConstraints(ctx, token, &rm, resolved)
+			if err != nil {
+				return nil, err
+			}
+		}
 
-// getModulesForService recursively fetches all modules for a given service manifest
-// this is done by iterating over all dependencies and then recursively calling ourself
-// to download their dependencies.
-func getModulesForService(ctx context.Context, sm *configuration.ServiceManifest,
-	deps []*configuration.TemplateRepository, modules map[string]*consideredModule) error {
-	for _, d := range deps {
-		if shouldSkipModule(modules, d) {
+		// check if the current module already is this version
+		// IDEA(jaredallard): In the future we probably want to see if
+		// the past version _also_ satisfies the new constraints, and if so
+		// skip re-resolving the module.
+		if version == resolved[importPath].Version {
+			modulesToResolve = modulesToResolve[1:]
 			continue
 		}
 
-		// create a module struct for this module, this resolves the latest version if
-		// the version wasn't set.
-		m, err := New(ctx, sm.Replacements[d.Name], d)
+		m, err := New(ctx, uri, &configuration.TemplateRepository{
+			Name:    importPath,
+			Channel: rm.conf.Channel,
+			Version: version,
+		})
 		if err != nil {
-			return errors.Wrapf(err, "failed to create dependency %q", d.Name)
+			return nil, err
 		}
 
-		// prefetch the manifest and FS
 		mf, err := m.Manifest(ctx)
 		if err != nil {
-			return errors.Wrapf(err, "failed to parse manifest %q", d.Name)
+			return nil, err
 		}
-		modules[d.Name] = &consideredModule{m, d.Prerelease}
 
-		// if we have dependencies, download those now
-		if len(mf.Modules) != 0 {
-			if err := getModulesForService(ctx, sm, mf.Modules, modules); err != nil {
-				return errors.Wrapf(err, "failed to process dependency of %q", d.Name)
-			}
+		// add the dependencies of this module to the stack to be resolved
+		for i := range mf.Modules {
+			modulesToResolve = append(modulesToResolve, resolveModule{
+				conf:   mf.Modules[i],
+				parent: importPath,
+			})
+		}
+
+		// set the module on our resolved module
+		resolved[importPath].Module = m
+
+		// resolve the next module
+		modulesToResolve = modulesToResolve[1:]
+	}
+
+	// convert the resolved modules to a list of modules
+	modules := make([]*Module, 0, len(resolved))
+	for _, m := range resolved {
+		modules = append(modules, m.Module)
+	}
+	return modules, nil
+}
+
+// getLatestModuleForConstraints returns the latest module that satisfies the provided constraints
+func getLatestModuleForConstraints(ctx context.Context, token cfg.SecretData,
+	m *resolveModule, resolved map[string]*resolvedModule) (string, error) {
+	constraints := resolved[m.conf.Name].constraints
+	if len(constraints) == 0 {
+		constraints = make([]constraint, 0)
+	}
+
+	// if we have a constraint, use it to resolve the version
+	if m.conf.Version != "" {
+		constraints = append(constraints, constraint{
```

Comment:
Thinking maybe even if we don't have a constraint we should still "commit it" as a `*` 🤔

Reply:
More thinking, instead of specifically tracking just constraints, I think we should log the entire tree of the resolution. That'll make some errors more clear, and get this as well (it's just type renaming I guess)